CNN workload

vb000 opened the pull request:
To maintain a standard interface to various workloads, and also to make it possible to tweak the interface in the future without too much manual work, I'd like to propose following project structure.

Each workload has 3 distinctive items:

1. Network architecture
2. Dataset and transformations
3. Training hyper-parameters

Apart from this, rest of the stuff can be common, and in fact, should be implemented common to make sure we'd have standard interface. To achieve this we could use python packaging to structure our workloads like below, and factor all the common stuff into `utils.py`:

```
pytorch-apps
├── __init__.py
├── lenet5
│   ├── inference.py
│   ├── __init__.py
│   ├── model.py
│   └── training.py
└── utils.py
```

Each workload would have `model.py`, `training.py` and `inference.py`. Of these three, `training.py` and `inference.py` are runnable scripts with `if __name__ == "__main__"` construct to perform respective task. And, of course, `model.py` would contain a class implementing network architecture.

To make sure each workload:
- takes standard command line arguments
- print standard messages during training and inference
- runs standard profiling tools

`utils.py` would contain routines like `argparse_inference`, `argparse_training`, `train` and `test`. All these need exact same implementation for all workload and hence could be factored into a importable module. For example, `lenet5/inference.py` can import `utils.py` and `model` using:

```
sys.path.append(os.path.join(os.path.dirname(__file__), os.pardir))

import model
import utils
```

This reduces the duplicate code and reduces the boilerplate needed when adding new workloads or changing the way profile, for instance. The main focus of this PR is to get this structure right, than making decisions on what precisely each module does, which is still open for discussion, I believe.

cbatten commented:
Lin and I had a discussion about how to organize workloads. Please do sync up with him. I imagine we could come up with something in between what he and I talked about and what you are proposing. A key thing is we need a structured way not just to run the full workload but also to run the key workloads with the reduced chunksizes ...

yodada commented on utils.py:
```diff
@@ -0,0 +1,97 @@
+                        help="Filename of the saved model")
+    return parser.parse_args()
+
+def argparse_training():
```
I really like the idea of having a common argparsing method. just wondering how should we handle workload specific options. e.g., `Rescyc` has a `validate` option.

Maybe we can have a common method called `add_common_options` (just like gem5's way if you are familiar)

vb000 replied:
I get your concern... instead of the common method doing `return parser.parse_args()`, it can do `return parser`.

Does this sounds good? We could add workload specific options to parser this way...

yodada replied:
I ended up doing it in a slightly different way. The common argument parsing method takes in a function which will injects workload specific options. so we can keep code like this in common space
https://github.com/cornell-brg/pytorch-apps/blob/2b7768a350acf635084e8414cbf33a856266d3d7/utils.py#L56-L59

yodada replied:
The expected behavior is: if either one is specified, the one not being specified will not run.

--training -> training only
--inference -> inference only
NONE -> both

vb000 replied:
Sounds good!

yodada commented on utils.py:
```diff
@@ -0,0 +1,97 @@
+    return parser.parse_args()
+
+# Train routine
+def train(model, loader, optimizer, loss_func, epochs, batches=None):
```
I guess it is also nice to have a common training / inference method ;)

yodada commented on lenet5/model.py:
```diff
@@ -0,0 +1,40 @@
+            nn.MaxPool2d(kernel_size=(2, 2), stride=2),
+            nn.Conv2d(16, 120, kernel_size=(5, 5)),
+        )
+
```
I think it is possible to merge these two Sequential blocks by having defining a layer that does the `view`

vb000 replied:
There's been lot of debate on this in pytorch forums ;). PyTorch officially recommended way is to not have separate `View` layer.

https://discuss.pytorch.org/t/non-legacy-view-module/131
https://discuss.pytorch.org/t/what-is-reshape-layer-in-pytorch/1110

vb000 replied:
I guess the class constructor should define the network architecture and `forward` defines the dataflow in the network. `View` is just a dataflow tweak and cannot be considered as part of the network architecture...

cbatten replied:
I think this is a good compromise. So our policy is the constructor defines the network architecture ... and forward can reshape inputs if necessary ...

vb000 commented:
> How about this: let me adapt your utils to create a method to populate common options
> So here are the options that I think should be in that common method
> 
> ```
>   --nepoch NEPOCH       number of training epochs (default: -1)
>   --nbatch NBATCH       number of training/inference batches (default: -1)
>   --batch_size BATCH_SIZE size of a batch (default: 32)
>   --hammerblade         run MLP MNIST on HammerBlade (default: False)
>   --training            run training phase (default: False)
>   --inference           run inference phase (default: False)
>   -v, --verbose         increase output verbosity (default: 0)
>   --save-model          save trained model to file (default: False)
>   --load-model          load trained model from file (default: False)
>   --model-filename MODEL_FILENAME
>                         filename of the saved model (default: trained_model)
>   --seed SEED           manual random seed (default: 42)
>   --dry                 dry run (default: False)
> ```
> 
> Any other sugguestions?

These options sound good to me. I expect this set to evolve over time though. So, the main idea is to factor it out a common module.

yodada commented:
Ok! I have finished refactoring mlp_mnist training and inference file. Could you take a look and see if it makes sense?

vb000 commented:
Great! Thanks for incorporating the changes, I'll update this workload on top of yours.

yodada commented:
So the <workload_name>_kernels.py file hosts the *isolated* kernels. These are kernels that account for most of the execution time. We also put the original data size and *reduced* (across pods) data size there 
https://github.com/cornell-brg/pytorch-apps/blob/bf05325f6ba73665c8ca8705258e678c03084c97/mlp_mnist/mlp_mnist_kernels.py#L116-L118

Of course there are many code that can be lifted to common space as well. Will work on it tomorrow

vb000 commented:
Just a minor thing, let's keep python indentation to 4 spaces? As per flake8 in hb-pytorch...

vb000 commented:
> So the <workload_name>_kernels.py file hosts the _isolated_ kernels. These are kernels that account for most of the execution time. We also put the original data size and _reduced_ (across pods) data size there
> https://github.com/cornell-brg/pytorch-apps/blob/bf05325f6ba73665c8ca8705258e678c03084c97/mlp_mnist/mlp_mnist_kernels.py#L116-L118
> 
> Of course there are many code that can be lifted to common space as well. Will work on it tomorrow

No worries, the plan sounds good!

yodada commented:
> Just a minor thing, let's keep python indentation to 4 spaces? As per flake8 in hb-pytorch...

Emmm what you are asking may not be possible XD

The fact that 1 tab = 2 spaces is hard coded in @cbatten ‘s gene

cbatten commented:
I am ok with four spaces in this context.

yodada commented:
@vb000 are you gonna change these tabs to 2 spaces? Just want to check before I touch them so we can avoid conflicts ;)

vb000 commented:
Those aren't tabs! All indents are 4 spaces...

yodada commented:
Yeah I meant "soft tabs".

vb000 commented:
I can change them in this pr to 4 spaces...

yodada commented:
I just enabled flake8 on this repo

vb000 commented:
- Adapted to new command line interface. 
- Implemented the model, training and inference in a single file.
- Fixed indentation in utils.py.
- `utils.py` and `lenet5` workload are flake8 clean.

yodada commented:
I added a template for workloads which has those (now) 7 sections. https://github.com/cornell-brg/pytorch-apps/blob/master/template/template.py

yodada commented:
Btw, I set flake8 to ignore E402

vb000 commented:
> I added a template for workloads which has those (now) 7 sections. https://github.com/cornell-brg/pytorch-apps/blob/master/template/template.py

Template looks great Lin! It really helps authors to maintain consistent style. Some thoughts:

- I think we should use `if __name__ == "__main__"` construct to be able to reuse the class we defined in the script in some other script. It's good to have that flexibility, and from what I understand, it's considered a good python practice to not run from the start of the script.
- I think clearly separating out model and hyper-parameters improves readability and makes the model easy to comprehend. I followed this approach of naming hyper-parameters in block letters:
https://github.com/cornell-brg/pytorch-apps/blob/5d8283b003baba3f657991b072a5848ec97fc822/lenet5/lenet5.py#L63-L73
As of now, these are things that define the model and training, so better to bunch them together?
- Hyper-parameters should be added as extra command line arguments by each workload. I agree we shouldn't move from defaults, but it might not hurt to have the flexibility. At least they show up in the help message for the user to understand what hyper-parameters are available.

What do you think?
